### PR TITLE
Simplify GPT-Neo local attention implementation

### LINKED ISSUE #11320: Irregular VRAM usage with gpt-neo inference with sequences longer than 250 tokens
## Environment info

- `transformers` version: 4.5.1 / HEAD
- Platform: Linux/Colab Pro
- Python version: 3.7
- PyTorch version (GPU?): 1.8.1 (CUDA 11.0)
- Tensorflow version (GPU?):
- Using GPU in script?: Yes, NVIDIA P100
- Using distributed or parallel set-up in script?:

### Who can help

@patil-suraj

## Information

Model I am using (Bert, XLNet ...): EleutherAI/gpt-neo-2.7B

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

## To reproduce

1. Install transformers in a Colab Pro notebook
2. Run this script to log peak memory usage for inference with increasing sequence length: https://gist.github.com/finetuneanon/7ce0ed5090a27a383abffbbbc0433a29
3. Wait for it to crash with an OOM error in the attention matmul somewhere above sequence length 1850

Output:
```
1870 5436434432
ok 6535669248
1871 5436434432
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-f2aeed4489bd> in <module>()
     21     return_dict_in_generate=True,
     22     repetition_penalty=1.2,
---> 23     pad_token_id=tokenizer.eos_token_id
     24   )
     25   del ids

13 frames
/usr/local/lib/python3.7/dist-packages/transformers/models/gpt_neo/modeling_gpt_neo.py in _attn(self, query, key, value, causal_mask, masked_bias, attn_dropout, attention_mask, head_mask)
    238         key = key.to(torch.float32)
    239 
--> 240         attn_weights = torch.matmul(query, key.transpose(-1, -2))
    241         attn_weights = torch.where(causal_mask, attn_weights, masked_bias.to(attn_weights.dtype))
    242 

RuntimeError: CUDA out of memory. Tried to allocate 4.59 GiB (GPU 0; 15.90 GiB total capacity; 9.75 GiB already allocated; 4.60 GiB free; 10.42 GiB reserved in total by PyTorch)
```

The full output can be found here: https://gist.github.com/finetuneanon/c7292ea676f57f5bb63803685d80bf5b

The output has the format:
```
sequence_length occupied_cuda_memory_before_inference
ok peak_occupied_cuda_memory_during_inference
```

Doing inference with real text has the same issue.

## Expected behavior

I expected memory usage to increase steadily instead of jumping around wildly, but I am not sure if this might actually be the correct behaviour. If it is correct, reliably doing inference on long sequences on 16GB of VRAM seems to be impossible, but sometimes it works.

I have also plotted the peak memory allocation during inference:

cudaoom

The green line is peak memory allocation, the brown line is the amount of memory in use before running inference.

### PR DESCRIPTION
**Pull request by finetunej**
# What does this PR do?

The current local self attention implementation for gpt-neo follows the implementation in mtf, which makes it hard to understand. This pull request implements local self attention by changing the bias mask in global attention. Local self attention is a sliding window where each token can only attend to the previous window_size tokens. This implementation clearly reflects this.

[Measured performance](https://gist.github.com/finetuneanon/5b2186c3555b652f387c86160cd89b55) (apply just 330686a3c0520c9727fe5ebed385e708d0178d77 and 269c497be1691556c830f61fa8f90001c692722f on #11320 patched to use) shows no noticable difference between implementations with respect to speed or VRAM usage. The results of both implementations are also identical.

Fixes #11320

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

Since this PR mostly removes code, no additional tests or documentation were written.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

gpt-neo: @patil-suraj

### CODE REVIEW

**Comment by finetuneanon** [comment]
I just found an issue with use_cache. Will fix it later.

**Comment by patil-suraj** [comment]
unstale

**Comment by finetuneanon** [comment]
I have used this version of the code quite extensively without issues and have found no difference in performance when doing evaluations. I have also just fixed the so far present warning about the local attention layer biases not being initialized from checkpoints.

**Comment by nostalgebraist** [comment]
> I have used this version of the code quite extensively without issues and have found no difference in performance when doing evaluations. I have also just fixed the so far present warning about the local attention layer biases not being initialized from checkpoints.

I want to second this.  For the past month, I have been [using](https://github.com/nostalgebraist/nostalgebraist-autoresponder/blob/ff2eafb1fbf6d670513f5d7b64ab99a739120945/stable_library_code/transformers/gpt_neo/modeling_gpt_neo.py) this version of the code in my bot -- which runs 24/7 and uses GPT-Neo 2.7B.  It's worked perfectly for me.

By comparison, the implementation currently on `master` [requires](https://github.com/nostalgebraist/nostalgebraist-autoresponder/blob/ff2eafb1fbf6d670513f5d7b64ab99a739120945/selector_model/selector_nn_neo.py#L45) counter-intuitive, inefficient manual padding to avoid OOM.

**Comment by patil-suraj** [comment]
Sorry about the super late response.

@finetuneanon  @nostalgebraist 
This sounds awesome! And yeah, I agree, the current implem is not super readable. 
Did you guys run any memory/speed benchmarks comparing the two versions? If so could you please post the results and the script so that we could take a look?

**Comment by finetuneanon** [comment]
Can provide those soon. Will close this PR though, because it's also included in #12106.

**Comment by finetuneanon** [comment]
Performance evaluation results are available here:
https://github.com/huggingface/transformers/pull/12106#discussion_r650008259

I am reopening this pull request due to the splitting of that PR.

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 1]
```diff
@@ -309,14 +309,6 @@ def forward(
             output_attentions=output_attentions,
         )
 
-        # cache the hidden_states instead of key_value_states
-        # for local attention layer
-        if self.attention_type == "local":
```
Comment (Express50) [inline src/transformers/models/gpt_neo/modeling_gpt_neo.py:508]: @finetuneanon I'm curious about this change. Does this mean the current implementation of use_cache is incorrect? This seems to fix the problem in #11787.

Comment (finetuneanon) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:508]: The current implementation of local attention is rather convoluted and requires use_cache to work differently for local attention layers. Think it's probably correct however. Due to the simplifications done in this PR, it is not necessary to handle local attention cache differently from regular cache anymore.

**Comment by StellaAthena** [comment]
@patil-suraj is there a reason this hasn't been merged yet? I notice that this method is used in #12493

**Comment by patil-suraj** [comment]
HI @finetuneanon, sorry to only come back to this now. 

This is a great solution, so let's go ahead with this!
The PR is in good shape already but needs some clean-up and we would also need to adapt tests.

Let me know if you want to continue working on this.

**Comment by finetuneanon** [comment]
Thanks for taking over.

**Pull request closed without merging**